codec-http/src/main/java/io/netty/handler/codec/http/websocketx/WebSocketCloseFrameHandler.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright 2019 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.codec.http.websocketx;
+
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelOutboundHandlerAdapter;
+import io.netty.channel.ChannelPromise;
+import io.netty.util.AttributeKey;
+import io.netty.util.internal.ObjectUtil;
+
+/**
+ * Send {@link CloseWebSocketFrame} message on channel close, if close frame was not sent before.
+ */
+final class WebSocketCloseFrameHandler extends ChannelOutboundHandlerAdapter {
+    private static final AttributeKey<Boolean> CLOSE_FRAME_SENT_ATTR_KEY = AttributeKey
+        .valueOf(WebSocketCloseFrameHandler.class.getName() + ".CLOSE_FRAME_SENT");
+
+    private final WebSocketCloseStatus closeStatus;
+
+    WebSocketCloseFrameHandler(WebSocketCloseStatus closeStatus) {
+        this.closeStatus = ObjectUtil.checkNotNull(closeStatus, "closeStatus");
+    }
+
+    @Override
+    public void close(final ChannelHandlerContext ctx, final ChannelPromise promise) throws Exception {
+        if (!closeFrameSent(ctx) && ctx.channel().isActive()) {
+            write(ctx, new CloseWebSocketFrame(closeStatus), ctx.voidPromise());
+        }
+        flush(ctx);
+        ctx.close(promise);
+    }
+
+    @Override
+    public void write(final ChannelHandlerContext ctx, Object msg, ChannelPromise promise) throws Exception {
+        if (closeFrameSent(ctx)) {
+            return;
```

Comment:
you should fail the promise in this case. and release the msg to ensure there are no leaks.

Reply:
fixed

Reply:
my fail. fixed it, but didn't push.